plugins/Validation/src/Validation.ts

```diff
@@ -1,223 +1,346 @@
+/* tslint:disable:no-any */
+import { State } from '@hookstate/core';
+import { PluginCallbacks, StateValueAtRoot } from '@hookstate/core/dist';
 
-import { Plugin, Path, StateValueAtPath, State } from '@hookstate/core';
+export const ValidationId = Symbol('Validation');
 
-export type ValidationSeverity = 'error' | 'warning';
+type ValidateFn<T> = (value: T) => boolean;
+type Path = readonly (string | number | symbol)[];
 
-export interface ValidationError {
-    readonly message: string;
-    readonly path: Path;
-    readonly severity: ValidationSeverity;
+interface NestedState {
+    state: State<any>;
+    parent?: NestedState;
```

Comment:
this relationship may cause large parent object being hold in RAM...

Reply:
NestedState is transient within the scope of a valid call.. this exists because I ran into issues reusing the State<atroot> and the re-render events firing off properly. potentially, this may not be needed the reverse tree walk can be eliminated in `conditionalValid`, by just doing a lookup each time, but will need to benchmark the performance hit.